frame/im-online/src/lib.rs

```diff
@@ -571,23 +571,46 @@ impl<T: Config> Pallet<T> {
 	pub(crate) fn send_heartbeats(
 		block_number: T::BlockNumber,
 	) -> OffchainResult<T, impl Iterator<Item = OffchainResult<T, ()>>> {
-		const HALF_SESSION: Percent = Percent::from_percent(50);
+		const START_HEARTBEAT_RANDOM_PERIOD: Permill = Permill::from_percent(10);
+		const START_HEARTBEAT_FINAL_PERIOD: Permill = Permill::from_percent(80);
```

Comment:
Shouldn't this be `HearbeatAfter::<T>::get()` or something like that?

100% of users will submit a heartbeat on the block after `HearbeatAfter::<T>::get()`, so we want this curve to top out at this point, and then we only catch the stragglers.

Reply:
`HeartbeatAfter` is only used when we don't get back a value from `NextSessionRotation::estimate_current_session_progress` which shouldn't happen with BABE (we can accurately track the session progress there). `HeartbeatAfter` is a static block number that's calculated when the session starts as being "half session", but a session progress is not linear with block numbers since a session is actually measured in slots and those may be skipped.